### PR TITLE
Implement guild sharding support

### LINKED ISSUE #12: Guild sharding support
Guild sharding, described here, has been deployed: https://github.com/hammerandchisel/discord-api-docs/issues/17

The library should expose this functionality.

### PR DESCRIPTION
**Pull request by ghost**
Implements guild sharding support as described by the Discord docs:

https://discordapp.com/developers/docs/topics/gateway#sharding

This change is backwards-compatible as a new method on the Discord struct is
used for connecting via sharding, `connect_via_shard`.

I tested this by using a bot in only two servers and running two instances:
- connecting via shard 0, replying to "!ping" with "0"
- connecting via shard 1, replying to "!ping" with "1"

Messages were always sent to the correct instance (as is to be expected), and DMs always went to instance 0 properly.

Closes #12.

(my editor also converted a 3-space indentation error into a tab, and although out-of-scope I figured I'd leave it in)

### CODE REVIEW

**Comment by SpaceManiac** [comment]
`shard_id` and `total_shards` are always either both None or both Some. What do you think about `shard_info: Option<(u8, u8)>`? Might make the code a little tidier.

File: src/connection.rs [hunk 1]
```diff
@@ -127,7 +130,9 @@ impl Connection {
 			ws_url: base_url.to_owned(),
 			token: token.to_owned(),
 			session_id: Some(session_id),
-			last_sequence: sequence;
+			last_sequence: sequence,
+            shard_id: shard_id,
+            total_shards: total_shards;
```
Comment (SpaceManiac) [inline src/connection.rs]: Pretty minor, but indentation error here

**Comment by ghost** [comment]
Sounds good. Not sure how that indentation error got there. Updated and rebased.

**Comment by SpaceManiac** [comment]
- When sharding is in use, `Event::Ready` has a `shard` field matching the one supplied while connecting. This should be handled.
- Is `u8` the right type here? Is there a 256-shard limit?
- I'm not 100% on `connect_via_shard`, but I'm not sure what'd be better. `connect_sharded`? Not a big deal.

**Comment by ghost** [comment]
There isn't a 256-shard limit (that's documented, at least), but there is going to likely be a 2500 max guild per shard limit. 2500 guilds per shard, it'd take someone being in 640,000 guilds to fulfill the limit. Not sure this will ever happen. Does u16 sound better?

**Comment by SpaceManiac** [comment]
Sounds reasonable, a `u8` is fine then. It can be changed later if it turns out to indeed be a limitation. If you add `shard` to `Event::Ready` (looks like `"shard": [0, 2]`) should be good to go.

**Comment by ghost** [comment]
Updated + rebased. I also modified `Connection` internals to use `Option<[u8; 2]>` instead of `Option<(u8, u8)>` since I think it makes more sense.

**Pull request merged**